### ISSUE TITLE
Handling MPI parallelism

### ISSUE THREAD

**Issue by samarjeet** [cs01/gdbgui#183]
##### Describe the issue
Can we use gdbgui with multiple processes which communicate via MPI. I generally use command line gdb and run multiple xterms windows with gdb. Each xterm windows runs one of the processes. 

##### Steps to replicate
On xterm 
 mpirun -n <# processes> xterm <executable_name> -e gdb  -x <gdb_commands_file>

##### Describe your environment
* Operating system and version: Linux
* gdbgui version (`gdbgui -v`):
* gdb version (`gdb -v`): 
* browser:
* python packages (`pip freeze`):

**Comment by cs01** [comment]
I have not used mpi before, so I read a little about it. It looks like it attempts to start multiple processes with the same command. This will not work with gdbgui because it will just attempt to run multiple servers on the same port (in parallel) which cannot be done.

However, if what you are trying to do is debug multiple programs simultaneously, mpi is not required for that; gdbgui spawns a new instance of gdb for each tab that you open, all running on the same port/server. You can get a high level view of how many gdb instances are running for your instance of gdbgui by going to gdbgui's menu then clicking Dashboard. But really all that needs to be done is open a new tab and view the gdb pid, and you'll see it's not the same as the first one that opened.

**Comment by YvanFournier** [comment]
Hello,

I'm not sure the best option is to comment on a closed issue or to reopen one, but am also interested in using gdbgui with MPI.

I can add a few specifications about MPI: MPI usually launches multiple processes, either on a single host or multiple ones, and usually provides slightly different environment variable values to each process so they can determine their own rank. Also, it must be mentioned that when restarting processes, the whole set of processes must be relaunched together (usually using an "mpiexec" or similar command), so processes cannot be restarted independently within a debugger).

I tried launching different gdbgui instances with different port numbers (before reading this issue), but it seems when 2 gdbgui instances are launched, the second is killed even when not using the same port number.

So using multiple tabs is intersesting, but in this case how can I connect to the correct process ? The issue is not the simply possibility os doing so (using gdb attach), but the possibility of automating it (so as to run and attach multiple processes frow a wrapper script).

For additional information, the case I am interested in is mainly debugging on localhost (for HPC environments with massive parallelism, having many windows or tabs become unworkable, so more specialized debuggers such as DDT or TotalView become necessary, though attaching to a few select processes remains very useful).

**Comment by cs01** [comment]
> I tried launching different gdbgui instances with different port numbers (before reading this issue), but it seems when 2 gdbgui instances are launched, the second is killed even when not using the same port number.

I cannot replicate this behavior. If you can reproduce this, please create a new issue as this is not intended.

> The issue is not the simply possibility os doing so (using gdb attach), but the possibility of automating it (so as to run and attach multiple processes frow a wrapper script).

I understand you need some sort of automation regarding pids, but I don't really understand it. Can you re-explain with MPI abstracted away? i.e. "I need to launch gdb then query an API to get the gdb process that is running."
